### PR TITLE
feat: prevent banning of connected base node in wallet

### PR DESCRIPTION
**Pull request by StriderDM**
Description
---
This PR excludes the connected base node in the wallet from being banned.

Motivation and Context
---
Usability

How Has This Been Tested?
---
cargo test --all
nvm use 12.22.6 && node_modules/.bin/cucumber-js --profile "ci" --tags "not @long-running and not @broken"

### CODE REVIEW

File: comms/dht/src/connectivity/mod.rs [hunk 1]
```diff
@@ -67,6 +70,7 @@ pub struct DhtConnectivity {
     config: DhtConfig,
     peer_manager: Arc<PeerManager>,
     node_identity: Arc<NodeIdentity>,
+    ban_exclusions: Arc<RwLock<Vec<NodeId>>>,
```
Comment (stringhandler) [inline comms/dht/src/connectivity/mod.rs]: Can this not just be local? I think the other Arc's are shared amongst multiple different structs, whereas there should only be one DhtConnectivity

```suggestion
    ban_exclusions: Vec<NodeId>,
```

File: comms/src/connectivity/requester.rs [hunk 2]
```diff
@@ -54,6 +54,7 @@ pub enum ConnectivityEvent {
     PeerConnected(PeerConnection),
     PeerConnectFailed(NodeId),
     PeerBanned(NodeId),
+    ExcludePeerBan(NodeId),
```
Comment (stringhandler) [inline comms/src/connectivity/requester.rs]: Is there something that subscribes to this event? I don't know if it's worth publishing it

**Review by stringhandler — COMMENTED** [review]
I would prefer the term "allow_list" and "add_peer_to_allow_list" rather than "exclude_ban_peer".

File: comms/src/connectivity/requester.rs (hunk 2 above)
Comment (philipr-za) [reply comms/src/connectivity/requester.rs]: It looks like this is being used to communicate to the DHT Connectivity Actor that doesn't have a direct handle because it doesn't provide a service API. However, I don't think this is really the ideal way to plumb this communication in. While it might make sense to add a peer to the Allowlist via the Comms::connectivity manager how do you manage that list afterwards in terms of seeing who is on it and removing peers from it?

**Review by philipr-za — COMMENTED** [review]
This functionality would typically be called an Allowlist (whitelist in the past).

I don't think it is ideal to make the interface for adding to the allowlist a call via the Comms::connectivity manager. I think you should add service handle to the Dht::connectivity actor to build out a service API. The reason is that you will not just be adding to the allowlist but also querying who is on it and removing peers from it.

This PR adds to the allowlist but doesn't remove the previous base node from the allow list when a new one is selected as described in the ticket.

File: comms/dht/src/connectivity/mod.rs [hunk 3]
```diff
@@ -67,6 +70,7 @@ pub struct DhtConnectivity {
     config: DhtConfig,
     peer_manager: Arc<PeerManager>,
     node_identity: Arc<NodeIdentity>,
+    message_rate_immune_node: Arc<RwLock<Option<NodeId>>>,
```
Comment (sdbondi) [inline comms/dht/src/connectivity/mod.rs]: Does not need to be locked - this state belongs to one task/service

File: comms/src/connectivity/manager.rs [hunk 4]
```diff
@@ -274,6 +274,11 @@ impl ConnectivityManagerActor {
                     error!(target: LOG_TARGET, "Error when banning peer: {:?}", err);
                 }
             },
+            MessageRateImmuneNode(node_id) => {
```
Comment (sdbondi) [inline comms/src/connectivity/manager.rs]: Should start with a verb
```suggestion
            AddMessageRateImmuneNode(node_id) => { // or AddPeerToAllowList in comms connectivity
```

**Review by sdbondi — COMMENTED** [review]
Not sure why the base node would be sending so many messages to the wallet - that doesn't seem right. 

Assuming there is a valid reason to add this, an allowlist of peers that cannot be banned can either be added to comms connectivity (responsible for banning peer's it is asked to - "don't ever ban these peers") or DHT connectivity (responsible for checking message rates - "don't ban this peer for message rate violation"). 

I think the former makes more sense, and there is already a service API to work with.

**Comment by philipr-za** [comment]
> Not sure why the base node would be sending so many messages to the wallet - that doesn't seem right.

Observed this occurring during a stress test and the base node was relaying SAF messages at a high rate.

**Comment by sdbondi** [comment]
> Observed this occurring during a stress test and the base node was relaying SAF messages at a high rate.

SAF messages are sent in bundles of messages, should be `num_messages/batch_Size` messages so seems excessive - probably the protocol could be improved by rather RPC streaming SAF messages - in general, I would prefer that to adding new exceptions/rules for edge cases but happy for something like this to go in for now.

**Comment by philipr-za** [comment]
> > Observed this occurring during a stress test and the base node was relaying SAF messages at a high rate.
> 
> SAF messages are sent in bundles of messages, should be `num_messages/batch_Size` messages so seems excessive - probably the protocol could be improved by rather RPC streaming SAF messages - in general, I would prefer that to adding new exceptions/rules for edge cases but happy for something like this to go in for now.

The sentiment is that there are several reasons that a wallet might ban a base node, with new ones being added in the future perhaps, but I would like to preserve the concept that a wallet will never ban the base node it has been explicitly set to use for any reason other than the WireMode byte. The WireMode byte though is enforced at a lower level I believe.

File: comms/dht/src/connectivity/mod.rs [hunk 5]
```diff
@@ -245,17 +247,24 @@ impl DhtConnectivity {
...
-                "Banning peer `{}` because of flooding. Message rate: {:.2}m/s", peer, mps
-            );
-            self.connectivity
-                .ban_peer_until(
-                    peer,
-                    self.config.ban_duration,
-                    "Exceeded maximum message rate".to_string(),
-                )
-                .await?;
+            if *self.message_rate_immune_node != Some(peer.clone()) {
```
Comment (philipr-za) [inline comms/dht/src/connectivity/mod.rs:250]: This should implemented as an allowlist in comms::connectivity

File: comms/src/connectivity/manager.rs [hunk 6]
```diff
@@ -748,6 +753,13 @@ impl ConnectivityManagerActor {
         }
         Ok(())
     }
+
+    async fn message_rate_immune_peer(&mut self, node_id: &NodeId) -> Result<(), ConnectivityError> {
+        info!(target: LOG_TARGET, "Excluding peer {} from banning", node_id);
+        self.publish_event(ConnectivityEvent::SetMessageRateImmuneNode(node_id.clone()));
```
Comment (philipr-za) [inline comms/src/connectivity/manager.rs:759]: Using an event to communicate with a service is not a good design.

File: comms/src/connectivity/requester.rs [hunk 7]
```diff
@@ -54,6 +54,7 @@ pub enum ConnectivityEvent {
     PeerConnected(PeerConnection),
     PeerConnectFailed(NodeId),
     PeerBanned(NodeId),
+    SetMessageRateImmuneNode(NodeId),
```
Comment (philipr-za) [inline comms/src/connectivity/requester.rs:57]: Rather than dealing with a specific case this should be an allowlist.

**Review by philipr-za — COMMENTED** [review]
This should not be implemented in dht::connectivity where you need to communicate with the dht actor via an event stream. Please implement this as an Allowlist where the Bans are processed: in the comms::connectivity service. Again, we are not focusing on a single ban case, the idea is implement an allow list that prevents any bans from occurring for a node on the allow list.

**Comment by StriderDM** [comment]
See PR #3642
